### PR TITLE
Add config file for openshift-ansible install

### PR DESCRIPTION
**Pull request by spadgett**
Needed for https://github.com/openshift/openshift-ansible/pull/8328

Adds gopkg.in/yaml.v2 dependency for YAML parsing (as a separate commit to make review easier)

### CODE REVIEW

**Comment by spadgett** [comment]
@ericchiang fyi

Do you know if I re-vendored correctly? The lines changed is much larger than I'd expect :(

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image has been pushed to Quay with tag `pr_65_build_293`

**Comment by ericchiang** [comment]
@spadgett you probably need to install https://github.com/sgotti/glide-vc

We should probably just switch to dep after this :)

**Comment by ericchiang** [comment]
Also I would _strongly_ recommend using https://github.com/ghodss/yaml in combination with [`encoding/json.Decoder.DisallowUnknownFields`](https://golang.org/pkg/encoding/json/#Decoder.DisallowUnknownFields) to make it an error to provide an unknown field. YAML parsing dropping incorrectly indented fields has been a huge issue for us in https://github.com/coreos/dex and I wish we had done that from the beginning.

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image failed.

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image failed.

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image failed.

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image failed.

**Comment by ggreer** [comment]
jenkins push

File: cmd/bridge/config.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+type Auth struct {
+	IssuerURL        string `yaml:"issuerURL"`
+	ClientID         string `yaml:"clientID"`
+	ClientSecretFile string `yaml:"clientSecretFile"`
+	LogoutPublicURL  string `yaml:"logoutPublicURL"`
+}
+
+// Customization holds configuration such as what logo to use.
+type Customization struct {
+	LogoImageName string `yaml:"logoImageName"`
```
Comment (ggreer) [inline cmd/bridge/config.go]: IMO we should drop this "feature". It's nothing but trouble.

**Comment by coreos-ui** [comment]
The PR image has been pushed to Quay with tag `pr_65_build_299`

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image has been pushed to Quay with tag `pr_65_build_386`

**Comment by spadgett** [comment]
@liggitt @deads2k are one of you able to give this an API review?

@jwforres fyi

**Comment by spadgett** [comment]
Example config:

```yaml
apiVersion: console.openshift.io/v1beta1
kind: ConsoleConfig
servingInfo:
  bindAddress: https://0.0.0.0:8443 
  certFile: /var/serving-cert/tls.crt
  keyFile: /var/serving-cert/tls.key
auth:
  clientID: openshift-console
  clientSecretFile: /var/oauth-config/clientSecret
  issuerURL: https://public-master.example.com:8443 
clusterInfo:
  consoleBaseAddress: https://console.router.default.svc.cluster.local
customization:
  documentationBaseURL: https://docs.openshift.org/3.11/ 
  logoImageName: origin
```

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image has been pushed to Quay with tag `pr_65_build_390`

**Comment by spadgett** [comment]
jenkins push

**Comment by spadgett** [comment]
@deads2k I added a separate `issuerCAFile`

**Comment by coreos-ui** [comment]
The PR image has been pushed to Quay with tag `pr_65_build_397`

File: cmd/bridge/config.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+// ClusterInfo holds information the about the cluster such as master public URL and console public URL.
+type ClusterInfo struct {
+	ConsoleBaseAddress string `yaml:"consoleBaseAddress"`
+	ConsoleBasePath    string `yaml:"consoleBasePath"`
+	MasterPublicURL    string `yaml:"masterPublicURL"`
+}
+
+// Auth holds configuration for authenticating with OpenShift. The auth method is assumed to be "openshift".
+type Auth struct {
+	IssuerURL        string `yaml:"issuerURL"`
```
Comment (enj) [inline cmd/bridge/config.go]: What is this used for?

File: cmd/bridge/config.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	IssuerCAFile     string `yaml:"issuerCAFile"`
```
Comment (enj) [inline cmd/bridge/config.go]: I assume we still have no solution for the routing problem on CI?

I assume we do *not* want to allow custom override of the OAuth metadata similar to the master:

```go
	// OAuthMetadataFile is a path to a file containing the discovery endpoint for OAuth 2.0 Authorization
	// Server Metadata for an external OAuth server.
	// See IETF Draft: // https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
	// This option is mutually exclusive with OAuthConfig
	OAuthMetadataFile string `json:"oauthMetadataFile"`
```

File: cmd/bridge/config.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	ClientID         string `yaml:"clientID"`
+	ClientSecretFile string `yaml:"clientSecretFile"`
```
Comment (enj) [inline cmd/bridge/config.go]: Does this need to support encryption on disk?

File: cmd/bridge/config.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	LogoutRedirect   string `yaml:"logoutRedirect"`
```
Comment (enj) [inline cmd/bridge/config.go]: Where is supposed to go "normally"?

File: cmd/bridge/config.go (hunk 2 above)
Comment (spadgett) [reply cmd/bridge/config.go]:
```
wellKnownURL := strings.TrimSuffix(c.issuerURL, "/") + "/.well-known/oauth-authorization-server"
```

https://github.com/openshift/console/blob/612f933e2cab16af2fdeecdb5cd33df0a62eb9b0/auth/auth_openshift.go#L31

File: cmd/bridge/config.go (hunk 5 above)
Comment (spadgett) [reply cmd/bridge/config.go]: Back to the console if no redirect set.

File: cmd/bridge/config.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+// ClusterInfo holds information the about the cluster such as master public URL and console public URL.
+type ClusterInfo struct {
+	ConsoleBaseAddress string `yaml:"consoleBaseAddress"`
+	ConsoleBasePath    string `yaml:"consoleBasePath"`
+	MasterPublicURL    string `yaml:"masterPublicURL"`
+}
+
+// Auth holds configuration for authenticating with OpenShift. The auth method is assumed to be "openshift".
+type Auth struct {
```
Comment (deads2k) [inline cmd/bridge/config.go]: I thought you were adding a CA bundle file to this.

File: cmd/bridge/config.go (hunk 2 above)
Comment (spadgett) [reply cmd/bridge/config.go]: OK, I plan to remove this and use the kubernetes service in-cluster and master public URL off-cluster for getting the well-known doc

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image failed.

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image failed.

**Comment by spadgett** [comment]
jenkins push

**Comment by spadgett** [comment]
OK, this is updated:

* Removed `issuerURL` from `auth` stanza (not needed for OpenShift OAuth discovery)
* Added `caFile` to `auth` stanza as a separate CA bundle file for the OAuth endpoints
* Updated to use k8s service (in-cluster) or master public URL (off-cluster) for OpenShift OAuth discovery

**Comment by coreos-ui** [comment]
The PR image has been pushed to Quay with tag `pr_65_build_400`

**Comment by spadgett** [comment]
Tested in-cluster and off-cluster on both Tectonic and OpenShift clusters.

File: cmd/bridge/config.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	ClientID         string `yaml:"clientID"`
+	ClientSecretFile string `yaml:"clientSecretFile"`
+	CAFile           string `yaml:"caFile"`
```
Comment (enj) [inline cmd/bridge/config.go]: I thought we wanted something like `oauthEndpointCAFile` to be specific.

Comment (spadgett) [reply cmd/bridge/config.go]: I don't mind the more specific name if you prefer.

**Comment by spadgett** [comment]
jenkins push

**Comment by coreos-ui** [comment]
The PR image has been pushed to Quay with tag `pr_65_build_401`

**Comment by spadgett** [comment]
Can I get an LGTM or further review comments? This is the last thing blocking the install changes from going in, and we're days from feature freeze.

**Comment by spadgett** [comment]
jenkins rebuild

**Comment by ericavonb** [comment]
@alecmerdler do you have time to give this a quick look over? are there any recent console changes that might affect this?

**Pull request merged**